tools/extract_challenges.py

```diff
@@ -0,0 +1,58 @@
+#!/usr/bin/env python
+
+'''Extract challenges from lessons.'''
+
+import sys
+import re
+
+
+USAGE = 'usage: extract_challenges [-o output_name] [filename...]'
+
+title_pat = re.compile(r'^>\s+.+\{.challenge\}')
+block_pat = re.compile(r'^>')
+
+def main(argv):
+    '''Main driver.'''
+
+    prog_name = argv[0]
+    filenames = argv[1:]
+    output_name = None
+    writer = sys.stdout
+
+    if filenames and (filenames[0] == '-o'):
+        if len(filenames) < 2:
+            fail(USAGE)
+        output_name = filenames[1]
+        filenames = filenames[2:]
+
+    if output_name:
+        writer = open(output_name, 'w')
+
+    if filenames:
+        for f in filenames:
+            with open(f, 'r') as reader:
+                extract(reader, sys.stdout)
+    else:
+        extract(sys.stdin, sys.stdout)
```

Comment:
When called without arguments, this script appears to hang; apparently, `for line in sys.stdin` never returns EOF. No usage message is printed.